Add a Painless Context REST API

jdconrad opened the pull request:
This PR adds an internal REST API for querying context information about Painless whitelists.

Commands include the following:
GET /_scripts/painless/_context -- retrieves a list of contexts
GET /_scripts/painless/_context?context=%name% retrieves all available information about the API for this specific context

In the future it should be easy to expand the API for external consumption or use with a web based IDE for retrieving info about a single type/single method, etc.

The follow up for this is to add a gradle task that can process these commands and generate whitelist API docs to keep them up-to-date.

jdconrad committed:
Add skeleton for context action.

jdconrad committed:
Store PainlessScriptEngine.t

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Add methods to access lookups for doc api.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Convert to use HandledRestAction

jdconrad committed:
Progress on API for returning methods

jdconrad committed:
Revent prettyprint

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Add a Painless Context REST API

elasticmachine commented:
Pinging @elastic/es-core-infra

jdconrad committed:
Fix style

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessContextAction.java:
```diff
@@ -0,0 +1,327 @@
+
+    public static class Request extends ActionRequest {
+
+        private String scriptContextName;
```
This fields needs to be serialized. This can be done by adding a constructor that accepts `StreamInput` and overwrite the `writeTo` method.

jdconrad replied:
Done.

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessContextAction.java:
```diff
@@ -0,0 +1,327 @@
+        }
+
+        public Response(StreamInput input) {
+            throw new UnsupportedOperationException();
```
We should have binary serialization here too. (implement this constructor and implement writeTo() method)

jdconrad replied:
Done.

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessContextAction.java:
```diff
@@ -0,0 +1,327 @@
+
+    public static class Response extends ActionResponse implements ToXContentObject {
+
+        private final PainlessScriptEngine painlessScriptEngine;
```
I think that `PainlessScriptEngine` should't be exposed in this response class. This something internal to painless itself. I think it makes more sense to introduce a simple pojo classes that are exposed here and that can serialize the information that users of this api need. The transport action can then built the response class with the pojo classes based on the painless script engine instance.

jdconrad replied:
I considered this and wasn't sure it was the right thing to do since so much information is being copied from the PainlessLookup.  But based on your feedback it makes sense for the serialization and encapsulation, so I'll work on changing it.  Thanks for this feedback!

jdconrad replied:
Added several POJOs to accommodate this.

martijnvg reviewed (commented):
The direction looks good. I left some comments about serialization.

jdconrad commented:
@martijnvg Thank you for the comments!  Will take a look soon.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Move to action package.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Add POJOs to handle serialization and response XContent.

jdconrad committed:
Update serialization.

jdconrad commented:
@martijnvg I think I've addressed all the serialization issues.  Would you please take another look when you get the chance?  Thanks in advance!

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Merge branch 'master' into docapi

martijnvg reviewed (commented):
This looks good @jdconrad. The only thing that is missing are some tests.
Can you add a test that extends `AbstractSerializingTestCase` which tests `PainlessContextInfo` class serialization wise? (this indirectly also tests the other pojos).
And maybe also add a yaml rest test for the painless context api?

jdconrad commented:
@martijnvg Thanks again for the feedback!  Will add the requested tests and get back to you.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Add serialization tests.

jdconrad committed:
Fix generics.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad commented:
@elasticmachine run elasticsearch-ci/1

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Add yaml tests.

jdconrad committed:
Add rest spec

jdconrad commented:
@martijnvg I added comprehensive serialization tests.  I also added a pair of yaml tests that execute against the two queries, but given the nature of the responses it was unrealistic to add checks against them at this time since the contains assertion will not work against lists.

jdconrad committed:
Fix naming convention test.

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextInfo.java:
```diff
@@ -0,0 +1,173 @@
+                painlessLookup.getClasses().stream().map(
+                        javaClass -> new PainlessContextClassInfo(javaClass, painlessLookup.lookupPainlessClass(javaClass))
+                ).collect(Collectors.toList()),
+                painlessLookup.getImportedPainlessMethodsKeys().stream().map(importedPainlessMethodKey -> {
```
maybe add `.sort(custom comparator)` to the chaining here in order to generate consistent response?
This will then also allow to the test the response in the yaml test?

jdconrad replied:
Sorted everything that made sense to sort.

martijnvg reviewed (commented):
> I also added a pair of yaml tests that execute against the two queries, but given the nature of the responses it was unrealistic to add checks against them at this time since the contains assertion will not work against lists.

See my comment in PainlessContextInfo constructor, maybe we can assert the response.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextInfo.java:
```diff
@@ -0,0 +1,173 @@
+    public static final ParseField NAME = new ParseField("name");
+    public static final ParseField CLASSES = new ParseField("classes");
+    public static final ParseField IMPORTED_METHODS = new ParseField("imported_methods");
+    public static final ParseField ClASS_BINDINGS = new ParseField("class_bindings");
```
Man, your eyes are way better than mine.  Fixed.

jdconrad commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextAction.java:
```diff
@@ -0,0 +1,230 @@
+
+    public static class TransportAction extends HandledTransportAction<Request, Response> {
+
+        PainlessScriptEngine painlessScriptEngine;
```
Fixed.

jdconrad committed:
Add imported info to class, add matches to yaml tests, and update
response object.

jdconrad commented:
@martijnvg So I changed the following:
1. Added a new field to PainlessContextClassInfo called imported that says if a class can be accessed through it's imported name (java.lang.Object  -> Object)
2. Simplified the response object for the action to have data structures more in line with what's actually being returned and sort the contexts upon return for easier testing.
3. Sorted constructors, methods, and fields with PainlessContextClassInfo.  Sorted the classes within PainlessContextInfo.
4. Added numerous matches against the yaml tests (thanks for the tip about lists again :) )

Ready for the next round of review.

jdconrad committed:
Fix ignored imports in ANTLR files.

jdconrad committed:
Merge branch 'master' into docapi

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextAction.java:
```diff
@@ -0,0 +1,251 @@
+        public Response(StreamInput in) throws IOException {
+            super(in);
+            scriptContextNames = in.readBoolean() ? in.readStringList() : null;
+            painlessContextInfo = in.readBoolean() ? new PainlessContextInfo(in) : null;
```
instead of the ternary expressions maybe use: `readOptionalWriteable(...)` and just use an empty list for scriptContextNames (if there isn't anything)?

jdconrad replied:
Thank you for readOptionalWritable!  That's what I was looking for, but failed to find it.

martijnvg commented on modules/lang-painless/src/main/java/org/elasticsearch/painless/action/PainlessContextAction.java:
```diff
@@ -0,0 +1,251 @@
+                out.writeStringCollection(scriptContextNames);
+            }
+
+            if (painlessContextInfo == null) {
```
same as the previous comment, maybe use `writeOptionalWriteable(...)` and empty list for `scriptContextNames `?

jdconrad replied:
Same as above.

martijnvg reviewed (approved):
LGTM - Thanks @jdconrad this looks great. 
I only left a minor comment.

jdconrad committed:
Merge branch 'master' into docapi

jdconrad committed:
Response to PR comments.

jdconrad commented:
@martijn Thank you for all the iterations of review!  Will commit as soon as CI passes.

martijnvg commented:
@jdconrad Looks great!
